app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala

```diff
@@ -147,6 +155,45 @@ class BooleanExprEval[F[_]: Monad] {
 }
 
 object BooleanExprEval {
+  def evalDuplicateExpr[T <: PageMode](
+    fields: Seq[FormCtx],
+    recData: RecData[SourceOrigin.Current]
+  ): Boolean = {
+    def canonicalStr(str: String, id: String): String = {
+      def lowerAndRemoveWhitespace(str: String): String = str.toLowerCase.filterNot(_.isWhitespace)
+      val fileUploadRegex: Regex = s"^[0-9]+_${id}_(.+)$$".r
+
+      fileUploadRegex.findFirstMatchIn(str) match {
+        case Some(res) => lowerAndRemoveWhitespace(res.group(1))
+        case _         => lowerAndRemoveWhitespace(str)
+      }
+    }
+
+    val filtered =
+      fields.flatMap(f => recData.variadicFormData.forBaseComponentIdLessThen(f.formComponentId.modelComponentId))
+    val compare = filtered
+      .map {
+        case (Pure(component), variadicValue) =>
+          (
+            component.maybeIndex,
+            component.baseComponentId.value,
+            None,
+            variadicValue.toSeq.map(canonicalStr(_, component.baseComponentId.value))
+          )
+        case (Atomic(component, atom), variadicValue) =>
+          (
+            component.maybeIndex,
+            component.baseComponentId.value,
+            Some(atom.value),
+            variadicValue.toSeq.map(canonicalStr(_, component.baseComponentId.value))
+          )
+      }
+      .groupBy(_._1)
+      .map(tuple => (tuple._1, tuple._2.map(tupleInner => (tupleInner._2, tupleInner._3, tupleInner._4)).toSet))
```

Comment:
Better not to overly use `_n` accessors. What about this alternative?
```suggestion
      .view
      .mapValues { values =>
        values.map { case (_, baseComponentId, atom, canonicals) => (baseComponentId, atom, canonicals) }.toSet
      }
```

Reply:
Have made the requested changes